drivers/nutdrv_qx.c

```diff
@@ -1008,6 +1008,100 @@ static int	fuji_command(const char *cmd, char *buf, size_t buflen)
 	return (int)strlen(buf);
 }
 
+/* SNR communication subdriver */
+static int	snr_command(const char *cmd, char *buf, size_t buflen)
+{
+	/*ATTENTION: This subdriver uses short buffer with length 102 byte*/
+	const struct {
+		const char	*str;	/* Megatec command */
+		const int	index;	/* String index for this command */
+		const char	prefix;	/* Character to replace the first byte in reply */
+	} command[] = {
+		{ "Q1\r", 0x03, '(' },
+		{ "F\r", 0x0d, '#' },
+		{ "I\r", 0x0c, '#' },
+		{ NULL, 0, '\0' }
+	};
+
+	int	i;
+
+	upsdebugx(3, "send: %.*s", (int)strcspn(cmd, "\r"), cmd);
+
+	for (i = 0; command[i].str; i++) {
+
+		int	retry;
+
+		if (strcmp(cmd, command[i].str)) {
+			continue;
+		}
+
+		for (retry = 0; retry < 10; retry++) {
+
+			int	ret;
+
+			ret = usb_get_string(udev, command[i].index, langid_fix, buf, 102);
```

Comment:
Thank you for taking a look at it!

Do you mean to add something like that before the first cycle ([line 1029](https://github.com/networkupstools/nut/pull/1008/files/b1b39d9b642a16696a831eec1699858439c7a866#diff-cb890ac7b82cb7a4f861284bcf39cc4f168312b61cb9455eb2755e345c0aa627R1029))?
```
	if (buflen < 102) {
		upsdebugx(4, "size of buf less than 102 byte!");
		return 0;
	}
```
I think it will protect against overflows if the buffer is too small.